templates/common/graph/graph.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"cmp"
+	"fmt"
+	"slices"
+
+	"golang.org/x/exp/maps"
+)
+
+// A DAG is a directed graph. It is list of nodes, identified by their slice
+// index, each having 0 or more outgoing edges to other node indices. So
+// [][]int{{1}, {}} is a graph with two nodes, where node 0 has an outgoing edge
+// to node 1, and node 1 has no outgoing edges.
+//
+// The edge directions are interpreted as "node index N depends on node indices
+// [X,Y,Z]". So the returned topological sort will place X, Y, and Z somewhere
+// before N.
+type DAG [][]int
+
+// ErrCyclic is returned when the provided graph has a cycle (so it can't be
+// topologically sorted).
+var ErrCyclic = fmt.Errorf("this directed graph has a cycle")
+
+type visitState int
+
+const (
+	unvisited visitState = iota
+	visiting
+	visited
+)
+
+// TopoSort is a topological (dependency-order) sort of the given node indices.
+// If there are multiple valid topological orderings, the choice of which one
+// you'll get back is deterministic but not controllable.
+//
+// The returned integers are indices into the input slice d. So a return value
+// of []int{3, 7, ...} means that d[3] has no dependencies and comes first in
+// the topological sort, and is followed by d[7], and so on.
+//
+// If any outbound edge index in any of the slices is out of range
+// 0<=edgeIndex<len(d) then it will panic.
+//
+// Sorting an empty graph returns an empty list.
+//
+// Note for maintainers: why didn't we import a third-party lib for this?
+// They're all either sketchy, unmaintainted, or require implementing lots of
+// cumbersome interfaces. And this entire algorithm is tiny.
+func TopoSort(d DAG) ([]int, error) {
```

Comment:
I'm a little confused about the decision to have a generic implementation and an int-based one? It seems a little odd. Can you explain?

Reply:
Wouldn't it be equally straightforward if you directly implement with the generic?

Reply:
Yeah, true. Maybe easier to read also. A generic sort using maps rather than slices would be bigger in memory and slower due to using generics and map lookups, but I suppose that doesn't really matter for this use case. I doubt anyone will install enough templates in a single repo for the difference in runtime between a int-slice and generic-map to matter (right?)

Reply:
Rewrote.